src/client.ts

```diff
@@ -19,103 +25,179 @@ export class APIClient {
+          statusText: error.statusText,
+        });
+      } else {
+        throw new IntegrationProviderAPIError({
+          cause: error,
+          endpoint,
+          status: error.status,
+          statusText: error.statusText,
+        });
+      }
     });
+  }
 
-    try {
-      await request;
-    } catch (err) {
-      throw new IntegrationProviderAuthenticationError({
-        cause: err,
-        endpoint: 'https://localhost/api/v1/some/endpoint?limit=1',
-        status: err.status,
-        statusText: err.statusText,
-      });
+  /**
+   * Iterates each corp resource in the provider.
+   *
+   * @param iteratee receives each resource to produce entities/relationship
+   */
+  public async iterateCorps(
+    iteratee: ResourceIteratee<SignalSciencesCorp>,
+  ): Promise<void> {
+    const endpoint = `${BASE_URI}/corps`;
+
+    const { data } = await this.getJsonWithRetries(endpoint, {
+      headers: this.generateHeaders(),
+    });
+
+    for (const corp of data) {
+      await iteratee(corp);
     }
   }
 
   /**
-   * Iterates each user resource in the provider.
+   * Iterates each user resource in the provider based on the provided corp.
    *
    * @param iteratee receives each resource to produce entities/relationships
    */
   public async iterateUsers(
-    iteratee: ResourceIteratee<AcmeUser>,
+    corpName: string,
+    iteratee: ResourceIteratee<SignalSciencesUser>,
   ): Promise<void> {
-    // TODO paginate an endpoint, invoke the iteratee with each record in the
-    // page
-    //
-    // The provider API will hopefully support pagination. Functions like this
-    // should maintain pagination state, and for each page, for each record in
-    // the page, invoke the `ResourceIteratee`. This will encourage a pattern
-    // where each resource is processed and dropped from memory.
-
-    const users: AcmeUser[] = [
-      {
-        id: 'acme-user-1',
-        name: 'User One',
-      },
-      {
-        id: 'acme-user-2',
-        name: 'User Two',
-      },
-    ];
-
-    for (const user of users) {
+    const endpoint = this.buildEndpoint(corpName, '/users');
+
+    const { data } = await this.getJsonWithRetries(endpoint, {
+      headers: this.generateHeaders(),
+    });
+
+    for (const user of data) {
       await iteratee(user);
     }
   }
 
+  private buildEndpoint(corp: string, path?: string) {
+    if (!corp) {
+      throw new IntegrationError({
+        message: `Parameter 'corp' is required, but received ${corp}`,
+        code: 'INVALID_PARAM_ERROR',
+      });
+    }
+
+    if (path) {
+      return `${BASE_URI}/corps/${corp}${path}`;
+    } else {
+      return `${BASE_URI}/corps/${corp}`;
+    }
+  }
+
+  private generateHeaders() {
+    const { apiUser, apiToken } = this.config;
+
+    return {
+      'x-api-user': apiUser,
+      'x-api-token': apiToken,
+      'Content-Type': 'application/json',
+    };
+  }
+
+  private async getJsonWithRetries(endpoint, options, retries = 0) {
```

Comment:
if we are trying to build a wrapper around `https`, it is going to get rather lengthy and for the most part could be "provider agnostic". That is to say, you could use it to make requests to any API. You may want to consider separating out this logic into its own `client.ts` file at the top-level, use generics, and then have a `signalSciences/client.ts` file that uses said `https` wrapper. Would help separate out the provider-specific logic out. This is done to some extent in the [graph-addigy integration](https://github.com/JupiterOne/graph-addigy/blob/main/src/addigy/client.ts).